File: src/lib/downloader/canonical.ts
/** URL identity, thumbnail filtering, original-size upgrades (gallery-dl / yt-dlp style). */

const TRACKING = new Set(["utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "fbclid", "gclid", "ref", "referrer"]);

export function decodeEntities(s: string): string {
  return s
    .replace(/&/g, "&")
    .replace(/"/g, '"')
    .replace(/&#39;|'/g, "'")
    .replace(/</g, "<")
    .replace(/>/g, ">")
    .replace(/&#(\d+);/g, (_, n) => String.fromCharCode(Number(n)))
    .replace(/&#x([0-9a-f]+);/gi, (_, n) => String.fromCharCode(parseInt(n, 16)));
}

export function absUrl(raw: string, base: string): string | null {
  try {
    const u = new URL(decodeEntities(raw.trim()), base);
    if (u.protocol !== "http:" && u.protocol !== "https:") return null;
    u.hash = "";
    return u.toString();
  } catch {
    return null;
  }
}

export function canonicalize(url: string): string {
  try {
    const u = new URL(url);
    u.hash = "";
    u.hostname = u.hostname.toLowerCase();
    for (const k of [...u.searchParams.keys()]) {
      if (TRACKING.has(k.toLowerCase())) u.searchParams.delete(k);
    }
    if (u.pathname.length > 1) u.pathname = u.pathname.replace(/\/+$/, "");
    return u.toString();
  } catch {
    return url;
  }
}

export function isThumbUrl(url: string): boolean {
  const path = url.split("?")[0]?.toLowerCase() ?? "";
  return /(?:^|[/_-])(thumb|thumbnail|thumbs|preview|small|icon|mini|tiny)(?:[/_-]|$)/.test(path) || /-\d{2,3}x\d{2,3}(?:\.\w+)?$/.test(path);
}

export function upgradeOriginal(url: string): string {
  try {
    const u = new URL(url);
    u.pathname = u.pathname
      .replace(/\/thumbs?\//gi, "/")
      .replace(/\/previews?\//gi, "/")
      .replace(/\/small\//gi, "/")
      .replace(/[-_](?:thumb|small|preview|150x150|300x300)/gi, "");
    for (const k of ["w", "h", "width", "height", "size"]) {
      const v = u.searchParams.get(k);
      if (v && Number(v) > 0 && Number(v) < 800) u.searchParams.delete(k);
    }
    return u.toString();
  } catch {
    return url;
  }
}

export function mediaKey(url: string): string {
  try {
    const u = new URL(canonicalize(url));
    const parts = u.pathname.replace(/\/+$/, "").toLowerCase().split("/").filter(Boolean);
    const last = parts[parts.length - 1] ?? "";
    if (/\.[a-z0-9]{2,5}$/.test(last)) {
      const stem = last
        .replace(/\.[a-z0-9]{2,5}$/, "")
        .replace(/[-_](?:thumb|small|preview|\d{2,4}x\d{2,4})$/i, "");
      return `${u.hostname}/${stem}`;
    }
    if (parts.length >= 3 && /^\d+$/.test(parts[parts.length - 1] ?? "") && /^\d+$/.test(parts[parts.length - 2] ?? "")) {
      return `${u.hostname}/${parts.slice(0, -2).join("/")}`;
    }
    return `${u.hostname}/${parts.join("/")}`;
  } catch {
    return url;
  }
}

export function looksLikeAlbum(url: string): boolean {
  return /\/(a|album|albums|gallery|galleries|g|post|posts|thread|threads|view|v|f|file|attachments?)\/[^/?#]+/i.test(url);
}

export function looksLikePagination(url: string): boolean {
  return /[?&]page=\d+/i.test(url) || /\/page-\d+/i.test(url) || /[?&]p=\d+/i.test(url) || /\/offset\/\d+/i.test(url);
}

export function inferredNextPage(current: string): string | null {
  try {
    const u = new URL(current);
    const q = u.searchParams.get("page") ?? u.searchParams.get("p");
    if (q && /^\d+$/.test(q)) {
      const key = u.searchParams.get("page") != null ? "page" : "p";
      u.searchParams.set(key, String(Number(q) + 1));
      return u.toString();
    }
    const m = u.pathname.match(/\/page-(\d+)\/?$/i);
    if (m) {
      u.pathname = u.pathname.replace(/page-\d+/i, `page-${Number(m[1]) + 1}`);
      return u.toString();
    }
    if (/\/threads\/[^/]+\/?$/i.test(u.pathname)) {
      u.pathname = u.pathname.replace(/\/?$/, "/") + "page-2";
      return u.toString();
    }
    return null;
  } catch {
    return null;
  }
}

export function bestSrcset(srcset: string): string | null {
  let bestUrl = "";
  let best = -1;
  for (const part of srcset.split(",")) {
    const bits = part.trim().split(/\s+/);
    const u = bits[0];
    if (!u) continue;
    const d = bits[1] ?? "1x";
    const n = d.endsWith("w") ? parseInt(d, 10) : d.endsWith("x") ? parseFloat(d) * 1000 : 1;
    if (n >= best) {
      best = n;
      bestUrl = u;
    }
  }
  return bestUrl || null;
}

File: desktop/src/SimpDownloader.Core/Canonical.cs
using System.Text.RegularExpressions;

namespace SimpDownloader.Core;

public static class Canonical
{
    static readonly HashSet<string> Tracking = new(StringComparer.OrdinalIgnoreCase)
    {
        "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "fbclid", "gclid", "ref", "referrer"
    };

    public static string DecodeEntities(string s) => s
        .Replace("\u0026amp;", "&")
        .Replace("\u0026quot;", "\"")
        .Replace("\u0026#39;", "'")
        .Replace("\u0026apos;", "'")
        .Replace("\u0026lt;", "<")
        .Replace("\u0026gt;", ">");

    public static string? Abs(string raw, string baseUrl)
    {
        raw = DecodeEntities(raw.Trim());
        if (!Uri.TryCreate(baseUrl, UriKind.Absolute, out var b)) return null;
        if (!Uri.TryCreate(b, raw, out var u)) return null;
        if (u.Scheme is not ("http" or "https")) return null;
        return u.GetLeftPart(UriPartial.Query);
    }

    public static Dictionary<string, string> Query(Uri u)
    {
        var d = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var part in u.Query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var kv = part.Split('=', 2);
            d[Uri.UnescapeDataString(kv[0])] = kv.Length > 1 ? Uri.UnescapeDataString(kv[1]) : "";
        }
        return d;
    }

    public static string Canonicalize(string url)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var u)) return url;
        var q = Query(u);
        foreach (var key in q.Keys.Where(Tracking.Contains).ToList()) q.Remove(key);
        var qs = string.Join("&", q.Select(kv => Uri.EscapeDataString(kv.Key) + "=" + Uri.EscapeDataString(kv.Value)));
        var path = u.AbsolutePath.Length > 1 ? u.AbsolutePath.TrimEnd('/') : u.AbsolutePath;
        var b = new UriBuilder(u) { Fragment = "", Host = u.Host.ToLowerInvariant(), Path = path, Query = qs };
        return b.Uri.ToString();
    }

    public static bool IsThumb(string url)
    {
        var path = url.Split('?')[0].ToLowerInvariant();
        return Regex.IsMatch(path, @"(?:^|[/_-])(thumb|thumbnail|thumbs|preview|small|icon|mini|tiny)(?:[/_-]|$)")
            || Regex.IsMatch(path, @"-\d{2,3}x\d{2,3}(?:\.\w+)?$");
    }

    public static string UpgradeOriginal(string url)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var u)) return url;
        var path = Regex.Replace(u.AbsolutePath, @"/thumbs?/", "/", RegexOptions.IgnoreCase);
        path = Regex.Replace(path, @"/previews?/", "/", RegexOptions.IgnoreCase);
        path = Regex.Replace(path, @"[-_](?:thumb|small|preview|150x150|300x300)", "", RegexOptions.IgnoreCase);
        return new UriBuilder(u) { Path = path }.Uri.ToString();
    }

    public static string MediaKey(string url)
    {
        try
        {
            var u = new Uri(Canonicalize(url));
            var parts = u.AbsolutePath.Trim('/').ToLowerInvariant().Split('/', StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length == 0) return u.Host;
            var last = parts[^1];
            if (Regex.IsMatch(last, @"\.[a-z0-9]{2,5}$"))
            {
                var stem = Regex.Replace(Path.GetFileNameWithoutExtension(last), @"[-_](?:thumb|small|preview|\d{2,4}x\d{2,4})$", "", RegexOptions.IgnoreCase);
                return u.Host + "/" + stem;
            }
            if (parts.Length >= 3 && parts[^1].All(char.IsDigit) && parts[^2].All(char.IsDigit))
                return u.Host + "/" + string.Join("/", parts.Take(parts.Length - 2));
            return u.Host + "/" + string.Join("/", parts);
        }
        catch { return url; }
    }

    public static bool LooksLikeAlbum(string url) =>
        Regex.IsMatch(url, @"/(a|album|albums|gallery|galleries|g|post|posts|thread|threads|view|v|f|file|attachments?)/[^/?#]+", RegexOptions.IgnoreCase);

    public static bool LooksLikePagination(string url) =>
        Regex.IsMatch(url, @"[?&]page=\d+", RegexOptions.IgnoreCase)
        || Regex.IsMatch(url, @"/page-\d+", RegexOptions.IgnoreCase)
        || Regex.IsMatch(url, @"[?&]p=\d+", RegexOptions.IgnoreCase);

    public static string? InferredNextPage(string current)
    {
        if (!Uri.TryCreate(current, UriKind.Absolute, out var u)) return null;
        var q = Query(u);
        foreach (var key in new[] { "page", "p" })
        {
            if (q.TryGetValue(key, out var vs) && int.TryParse(vs, out var n))
            {
                q[key] = (n + 1).ToString();
                var qs = string.Join("&", q.Select(kv => Uri.EscapeDataString(kv.Key) + "=" + Uri.EscapeDataString(kv.Value)));
                return new UriBuilder(u) { Query = qs }.Uri.ToString();
            }
        }
        var m = Regex.Match(u.AbsolutePath, @"/page-(\d+)/?$", RegexOptions.IgnoreCase);
        if (m.Success)
            return new UriBuilder(u) { Path = Regex.Replace(u.AbsolutePath, @"page-\d+", "page-" + (int.Parse(m.Groups[1].Value) + 1), RegexOptions.IgnoreCase) }.Uri.ToString();
        if (Regex.IsMatch(u.AbsolutePath, @"/threads/[^/]+/?$", RegexOptions.IgnoreCase))
            return new UriBuilder(u) { Path = u.AbsolutePath.TrimEnd('/') + "/page-2" }.Uri.ToString();
        return null;
    }

    public static string? BestSrcset(string srcset)
    {
        string best = "";
        var score = -1.0;
        foreach (var part in srcset.Split(','))
        {
            var bits = part.Trim().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
            if (bits.Length == 0) continue;
            var d = bits.Length > 1 ? bits[1] : "1x";
            double n = 1;
            if (d.EndsWith("w") && double.TryParse(d.TrimEnd('w'), out var w)) n = w;
            else if (d.EndsWith("x") && double.TryParse(d.TrimEnd('x'), out var x)) n = x * 1000;
            if (n >= score) { score = n; best = bits[0]; }
        }
        return string.IsNullOrEmpty(best) ? null : best;
    }
}
